network/port/port.go

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Genome Research Ltd.
+ *
+ * Author: Sendu Bala <sb10@sanger.ac.uk>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ ******************************************************************************/
+
+package port
+
+import (
+	"fmt"
+	"net"
+)
+
+const maxPort = 65535
+const maxTries = maxPort - 10000
+
+// listener interface is used instead of *net.TCPListener directly, so that we
+// can test with a mock version.
+type listener interface {
+	Addr() net.Addr
+	Close() error
+}
+
+// Checker is used to check for available ports on a host.
+type Checker struct {
+	addr      *net.TCPAddr
+	listeners []listener
+	ports     map[int]bool
+}
+
+// NewChecker returns a Checker that can check ports on the given host.
+func NewChecker(host string) (*Checker, error) {
+	addr, err := net.ResolveTCPAddr("tcp", host+":0")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Checker{
+		addr:  addr,
+		ports: make(map[int]bool),
+	}, nil
+}
+
+// AvailableRange asks the operating system for ports until it is given a
+// contiguous range of ports size long. It returns the first and last of those
+// port numbers, having released all the ports it took, so they are ready for
+// you to use.
+//
+// NB: there is the potential for a race condition here, where once released,
+// another process gets one of the ports before you use it, so start listening
+// on all the returned ports as soon as possible after calling this.
+func (c *Checker) AvailableRange(size int) (int, int, error) {
+	var err error
+
+	defer func() {
+		err = c.release(err)
+	}()
+
+	var min, max int
+
+	for i := 0; i < maxTries; i++ {
+		var port int
+
+		port, err = c.availablePort()
+		if err != nil {
```

Comment:
Why are we breaking the loop here? 
Previously it was alright with "continue", it needs to retry for maxTries times, right?

Reply:
Now we give up on the first failure, instead of combining lots of failures, because as per this test case, we can get ~65k failures and run out of memory!